### PR TITLE
SI-8589 Performance improvement for ArrayCharSequence.toString

### PR DESCRIPTION
**Pull request by jeroentervoorde**
While using a csv parser to process large char arrays i ran into the problem that the toString method of ArrayCharSequence was very slow on large buffers (and, as it appears, also on small buffers)
The problematic method:
override def toString = xs drop start take length mkString ""

I created a faster version which is already almost 50% faster on an empty charsequence and only becomes relatively faster when the array gets bigger.

See https://issues.scala-lang.org/browse/SI-8589

### CODE REVIEW

**Comment by retronym** [comment]
Review by @Ichoran

File: src/library/scala/runtime/SeqCharSequence.scala [hunk 1]
```diff
@@ -41,5 +41,5 @@ final class ArrayCharSequence(val xs: Array[Char], start: Int, end: Int) extends
       new ArrayCharSequence(xs, start1, start1 + newlen)
     }
   }
-  override def toString = xs drop start take length mkString ""
+  override def toString = if (start >= end) "" else new String(xs, start, length)
```
Comment (Ichoran) [inline src/library/scala/runtime/SeqCharSequence.scala]: This has different behavior when the bounds are incorrect.  The old version would return an empty string.  The new one throws StringIndexOutOfBoundsException.  Unless there's a compelling reason to prefer the latter, you should check the bounds.  (The class itself does not.)  (In particular, start >= 0, start+length <= s.length.)

**Comment by jeroentervoorde** [comment]
Thanks for the review, and sorry for missing that. I assumed the bounds would be ok. 
I've changed it to this:

  override def toString = {
    val start = Math.max(this.start, 0)
    val end = Math.min(xs.length, start + length)

```
if (start >= end) "" else new String(xs, start, end - start)
```

  }

And tested that with this:

  val chars = "jdfhfjhdfkjasdfjkdsajkfasdjkfhasdkjfhsdafkj48r848rweruqewiuroqweoiruiwqeruoiwe".toCharArray

  val ok = (1 to 10000000).forall { _ =>
    val sign1 = if (Random.nextBoolean()) 1 else -1
    val sign2 = if (Random.nextBoolean()) 1 else -1
    val i1 = Random.nextInt(100) \* sign1
    val i2 = Random.nextInt(100) \* sign2

```
val s1 = new FastArrayCharSequence(chars, i1, i2).toString()
val s2 = new ArrayCharSequence(chars, i1, i2).toString()
s1 == s2
```

  }
  println(ok)

I am a bit new to this, can i just commit the change to my branch (i did already) or do i need to rebase it into a single commit?

**Comment by Ichoran** [comment]
It is preferable to rebase into a single commit (git rebase -i HEAD~2 should do the trick).

**Comment by adriaanm** [comment]
ping @Ichoran

**Comment by jeroentervoorde** [comment]
I tried rebasing it but i had to merge my remote branch again creating an extra empty commit. Not sure why... I hope it is ok like this.

File: src/library/scala/runtime/SeqCharSequence.scala [hunk 2]
```diff
@@ -41,5 +41,10 @@ final class ArrayCharSequence(val xs: Array[Char], start: Int, end: Int) extends
       new ArrayCharSequence(xs, start1, start1 + newlen)
     }
   }
-  override def toString = xs drop start take length mkString ""
+  override def toString = {
+    val start = Math.max(this.start, 0)
```
Comment (Ichoran) [inline src/library/scala/runtime/SeqCharSequence.scala]: Use math not Math

**Comment by Ichoran** [comment]
@jeroentervoorde - You should be able to get a clean rebase by following these steps, assuming you've got "origin" as your remote github repository and "scala" as the main Scala one (let me know if this doesn't make sense to you; note that we normally name the branches issue/8589 not SI_8589):

(1) checkout your 2.11.x branch:  git checkout 2.11.x
(2) pull the most recent Scala 2.11.x: git pull scala 2.11.x
(3) sync your github branch: git push origin 2.11.x
(4) switch to your new branch: git checkout SI_8589
(5) rebase to 2.11.x: git rebase 2.11.x
(6) merge all the existing patches: 
  (i) git rebase -i HEAD~4
  (ii) change "pick" to "squash" for every patch of yours past the first
(7) push to github, forcing an overwrite of inconsistent history: git push -f origin SI_8589

**Comment by jeroentervoorde** [comment]
I think it is ok now. Thanks very much for your help. I'm off reading the git book again :)

**Comment by Ichoran** [comment]
That part looks good.  Any chance you can do it one more time to use `math` (i.e. `scala.math`) instead of `Math` (i.e. `java.lang.Math`), as I mentioned before?

**Comment by jeroentervoorde** [comment]
Sorry, i missed that. It's updated now.

**Comment by adriaanm** [comment]
ping @Ichoran

**Comment by Ichoran** [comment]
@adriaanm - Whoops, overlooked this, sorry! LGTM

**Comment by adriaanm** [comment]
No worries -- please start your comment with LGTM when the LGTM is final -- that's what triggers the bot to add the "reviewed" label.

**Comment by adriaanm** [comment]
Thanks for the improvement, @jeroentervoorde!

**Pull request merged**